src/main/java/org/truffleruby/core/encoding/EncodingNodes.java

```diff
@@ -66,22 +68,59 @@ protected boolean isAsciiCompatible(RubyEncoding encoding) {
         }
     }
 
+    // MRI: enc_compatible_str and enc_compatible_latter
+    @ImportStatic({ CodeRange.class, RopeGuards.class })
     public abstract static class NegotiateCompatibleRopeEncodingNode extends RubyBaseNode {
 
+        @Child RopeNodes.CodeRangeNode codeRangeNode;
+
         public abstract RubyEncoding executeNegotiate(Rope first, RubyEncoding firstEncoding, Rope second,
                 RubyEncoding secondEncoding);
 
+        public RubyEncoding negotiate(RopeWithEncoding first, RopeWithEncoding second) {
+            return executeNegotiate(first.getRope(), first.getEncoding(), second.getRope(), second.getEncoding());
+        }
+
         public static NegotiateCompatibleRopeEncodingNode create() {
             return NegotiateCompatibleRopeEncodingNodeGen.create();
         }
 
-        @Specialization(guards = "firstEncoding == secondEncoding")
+        @Specialization(guards = {
+                "firstEncoding == cachedEncoding",
+                "secondEncoding == cachedEncoding",
+        }, limit = "getCacheLimit()")
+        protected RubyEncoding negotiateSameEncodingCached(
+                Rope first, RubyEncoding firstEncoding, Rope second, RubyEncoding secondEncoding,
+                @Cached("firstEncoding") RubyEncoding cachedEncoding) {
+            assert first.encoding == firstEncoding.jcoding && second.encoding == secondEncoding.jcoding;
+            return cachedEncoding;
+        }
+
+
+        @Specialization(guards = "firstEncoding == secondEncoding", replaces = "negotiateSameEncodingCached")
         protected RubyEncoding negotiateSameEncodingUncached(
                 Rope first, RubyEncoding firstEncoding, Rope second, RubyEncoding secondEncoding) {
             assert first.encoding == firstEncoding.jcoding && second.encoding == secondEncoding.jcoding;
             return firstEncoding;
         }
 
+        @Specialization(guards = {
+                "firstEncoding != secondEncoding",
+                "firstEncoding == cachedEncoding",
+                "isStandardEncoding(firstEncoding)",
```

Comment:
Should be `isStandardEncoding(cachedEncoding)` so it's only checked once and not on the fast path, I'll do that.